instrumentation/dubbo/dubbo-apache-2.7/testing/src/main/groovy/io/opentelemetry/instrumentation/dubbo/apache/v2_7/AbstractDubboTest.groovy

```diff
@@ -0,0 +1,155 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.dubbo.apache.v2_7
+
+import io.opentelemetry.instrumentation.dubbo.apache.v2_7.api.HelloService
+import io.opentelemetry.instrumentation.dubbo.apache.v2_7.impl.HelloServiceImpl
+import io.opentelemetry.instrumentation.test.InstrumentationSpecification
+import io.opentelemetry.instrumentation.test.utils.PortUtils
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes
+import org.apache.dubbo.config.ApplicationConfig
+import org.apache.dubbo.config.ProtocolConfig
+import org.apache.dubbo.config.ReferenceConfig
+import org.apache.dubbo.config.RegistryConfig
+import org.apache.dubbo.config.ServiceConfig
+import org.apache.dubbo.config.bootstrap.DubboBootstrap
+import org.apache.dubbo.config.utils.ReferenceConfigCache
+import org.apache.dubbo.rpc.service.GenericService
+import spock.lang.Unroll
+
+import static io.opentelemetry.api.trace.Span.Kind.CLIENT
+import static io.opentelemetry.api.trace.Span.Kind.SERVER
+import static io.opentelemetry.instrumentation.test.utils.TraceUtils.basicSpan
+import static io.opentelemetry.instrumentation.test.utils.TraceUtils.runUnderTrace
+
+@Unroll
+abstract class AbstractDubboTest extends InstrumentationSpecification {
+
+  ReferenceConfig<HelloService> configureClient(int port) {
+    ReferenceConfig<HelloService> reference = new ReferenceConfig<>()
+    reference.setInterface(HelloService)
+    reference.setGeneric("true")
+    reference.setUrl("dubbo://localhost:" + port)
+    return reference
+  }
+
+  ServiceConfig configureServer() {
+    def registerConfig = new RegistryConfig()
+    registerConfig.setAddress("N/A")
+    ServiceConfig<HelloServiceImpl> service = new ServiceConfig<>()
+    service.setInterface(HelloService)
+    service.setRef(new HelloServiceImpl())
+    service.setRegistry(registerConfig)
+    return service
+  }
+
+  def "test apache dubbo base #dubbo"() {
+    setup:
+    def port = PortUtils.randomOpenPort()
+    def protocolConfig = new ProtocolConfig()
+    protocolConfig.setPort(port)
+
+    DubboBootstrap bootstrap = DubboBootstrap.getInstance()
+    bootstrap.application(new ApplicationConfig("dubbo-test-provider"))
+      .service(configureServer())
+      .protocol(protocolConfig)
+      .start()
+
+    def consumerProtocolConfig = new ProtocolConfig()
+    consumerProtocolConfig.setRegister(false)
+
+    def reference = configureClient(port)
+    DubboBootstrap consumerBootstrap = DubboBootstrap.getInstance()
+    consumerBootstrap.application(new ApplicationConfig("dubbo-demo-api-consumer"))
+      .reference(reference)
+      .protocol(consumerProtocolConfig)
+      .start()
+
+    when:
+    GenericService genericService = ReferenceConfigCache.getCache().get(reference) as GenericService
+    def o = new Object[1]
+    o[0] = "hello"
+    def response = runUnderTrace("parent") {
+      genericService.$invoke("hello", [String.getName()] as String[], o)
+    }
+
+    and:
```

Comment:
I'm still not a spock expert but I think the general pattern would be

```
when:
then:
and:
when:
then:
```

Or it may make sense to separate sync and async. Currently the assertions being in one big block make it hard to understand.

Reply:
Yes, I will separate them.